code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/internal/EdgePersonalizationResponseHandler.java

```diff
@@ -163,128 +199,300 @@ void fetchMessages(final List<Surface> surfaces) {
+                final List<PropositionItem> propositionItemList = new ArrayList<>();
+                propositionItemList.add(propositionItem);
+                final Proposition proposition = new Proposition(propositionInfo.id, propositionInfo.scope, propositionInfo.scopeDetails, propositionItemList);
+                inboundPropositionList.add(proposition);
+            }
+            propositionMap.put(surface, inboundPropositionList);
+        }
+        return propositionMap;
+    }
+
+    /**
+     * Attempts to parse any in-app message or message feed rules contained in the provided {@code List<Proposition>}. Any valid rule {@code JSONObject}s
+     * found will be returned in a {code List<LaunchRule>}.
+     *
+     * @param propositions     A {@link List<Proposition>} propositions to be processed
+     * @param expectedSurfaces A {@link List<Surface>} containing the expected app surfaces
+     * @param clearExisting    {@code boolean} if true the existing cached propositions are cleared and new message rules are replaced in the {@code LaunchRulesEngine}
+     * @param persistChanges   {@code boolean} if true the passed in {@code List<Proposition>} are added to the cache
+     * @return {@link Map<InboundType, List<LaunchRule>>} containing a mapping between {@link InboundType} and {@link LaunchRule}s parsed from the passed in propositions
+     */
+    private Map<InboundType, List<LaunchRule>> parsePropositions(final List<Proposition> propositions, final List<Surface> expectedSurfaces, final boolean clearExisting, final boolean persistChanges) {
+        final Map<InboundType, List<LaunchRule>> rules = new HashMap<>();
         final Map<String, PropositionInfo> tempPropositionInfo = new HashMap<>();
+        Map<Surface, List<Proposition>> tempPropositions = new HashMap<>();
+        Map<Surface, List<Proposition>> inAppPropositions = new HashMap<>();
+
+        if (clearExisting) {
+            clearSurfaces(expectedSurfaces);
+        }
 
-        if (propositions != null && !propositions.isEmpty()) {
-            for (final PropositionPayload proposition : propositions) {
-                if (proposition.propositionInfo != null && !expectedScope.equals(proposition.propositionInfo.scope)) {
-                    Log.debug(MessagingConstants.LOG_TAG, SELF_TAG, "processPropositions - Ignoring proposition where scope (%s) does not match expected scope (%s).", proposition.propositionInfo.scope, expectedScope);
+        if (MessagingUtils.isNullOrEmpty(propositions)) {
+            return rules;
+        }
+
+        String inboundTypeString;
+        InboundType inboundType = null;
+        List<LaunchRule> parsedRules;
+        for (final Proposition proposition : propositions) {
+            final String scope = proposition.getScope();
+            for (final Surface surface : expectedSurfaces) {
```

Comment:
This is supposed to be a local loop that needs to terminate soon after checking whether the provided `scope` matches one of the values in the `expectedSurfaces` list.